File: scripts/capture_storyboard.py
#!/usr/bin/env python3
"""Capture ResearchOS UI states for Figma storyboard handoff."""
from __future__ import annotations

import json
import os
import sys
import time
from datetime import datetime, timezone
from pathlib import Path

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
OUT_DIR = ROOT / ".figma-storyboard"
BASE = os.environ.get("KOI_UI_URL", "http://127.0.0.1:8080")
VIEWPORT = {"width": 1440, "height": 900}


def wait_for_app(page) -> None:
    page.goto(f"{BASE}/", wait_until="networkidle")
    page.wait_for_selector("#project-list .project-list__item, #project-list button", timeout=60_000)
    page.wait_for_timeout(1500)


def show_modal(page, modal_id: str) -> None:
    page.evaluate(
        """(id) => {
      document.querySelectorAll('.modal').forEach(m => m.classList.add('hidden'));
      document.getElementById('agent-chat-panel')?.classList.add('hidden');
      document.getElementById('rq-bell-panel')?.classList.add('hidden');
      document.getElementById(id)?.classList.remove('hidden');
    }""",
        modal_id,
    )
    page.wait_for_timeout(400)


def open_panel(page, panel_id: str) -> None:
    page.evaluate(
        """(id) => {
      document.getElementById(id)?.classList.remove('hidden');
    }""",
        panel_id,
    )
    page.wait_for_timeout(400)


def reset_ui(page) -> None:
    page.evaluate(
        """() => {
      document.querySelectorAll('.modal').forEach(m => m.classList.add('hidden'));
      document.getElementById('agent-chat-panel')?.classList.add('hidden');
      document.getElementById('rq-bell-panel')?.classList.add('hidden');
      document.getElementById('projects-sidebar')?.classList.remove('is-open');
      const toggle = document.getElementById('btn-projects-toggle');
      if (toggle) toggle.setAttribute('aria-expanded', 'false');
    }"""
    )
    page.wait_for_timeout(300)


def set_theme(page, theme: str) -> None:
    page.evaluate(
        """(t) => {
      localStorage.setItem('koi-theme', t);
      document.documentElement.setAttribute('data-theme', t);
    }""",
        theme,
    )
    page.wait_for_timeout(200)


def set_view_mode(page, mode: str) -> None:
    page.select_option("#view-mode-select", mode)
    page.wait_for_timeout(1200)


def shot(page, name: str) -> str:
    file = OUT_DIR / f"{name}.png"
    page.screenshot(path=str(file), full_page=False)
    return str(file)


def open_sidebar(page) -> None:
    page.evaluate("""() => {
      const sidebar = document.getElementById('projects-sidebar');
      const toggle = document.getElementById('btn-projects-toggle');
      if (sidebar) sidebar.classList.add('is-open');
      if (toggle) toggle.setAttribute('aria-expanded', 'true');
    }""")
    page.wait_for_timeout(400)


def pick_project_with_data(page) -> None:
    llm = page.locator("#project-list button", has_text="Проблема обучения LLM")
    if llm.count():
        btn = llm.first
        if btn.get_attribute("aria-current") == "true":
            page.wait_for_timeout(400)
            return
        btn.click(force=True)
    else:
        page.locator("#project-list button[data-project-id]").first.click(force=True)
    page.wait_for_timeout(1200)


def click_first_method_kanban(page) -> bool:
    clicked = page.evaluate(
        """() => {
      const el = document.querySelector('.node-kanban-compact, .node-kanban-below');
      if (!el) return false;
      el.dispatchEvent(new MouseEvent('click', { bubbles: true, cancelable: true }));
      return true;
    }"""
    )
    if clicked:
        page.wait_for_timeout(1000)
        return page.locator("#kanban-modal:not(.hidden)").count() > 0
    return False


def capture_theme(page, theme: str) -> list[str]:
    prefix = "light" if theme == "light" else "dark"
    files: list[str] = []

    wait_for_app(page)
    set_theme(page, theme)
    page.reload(wait_until="networkidle")
    page.wait_for_timeout(1500)
    reset_ui(page)
    pick_project_with_data(page)

    for mode in ("chief", "teamlead", "researcher"):
        reset_ui(page)
        set_view_mode(page, mode)
        files.append(shot(page, f"{prefix}-01-lab-{mode}"))

    reset_ui(page)
    set_view_mode(page, "chief")
    open_sidebar(page)
    files.append(shot(page, f"{prefix}-02-sidebar-open"))

    reset_ui(page)
    set_view_mode(page, "chief")
    open_panel(page, "agent-chat-panel")
    files.append(shot(page, f"{prefix}-03-agent-chat"))

    reset_ui(page)
    open_panel(page, "rq-bell-panel")
    page.evaluate("""() => {
      const p = document.getElementById('rq-bell-panel');
      if (p) { p.hidden = false; p.classList.remove('hidden'); }
    }""")
    files.append(shot(page, f"{prefix}-04-rq-bell"))

    reset_ui(page)
    show_modal(page, "settings-modal")
    files.append(shot(page, f"{prefix}-05-settings"))

    reset_ui(page)
    show_modal(page, "create-project-modal")
    files.append(shot(page, f"{prefix}-06-create-project"))

    reset_ui(page)
    set_view_mode(page, "researcher")
    show_modal(page, "node-modal")
    page.evaluate("""() => {
      const t = document.getElementById('node-modal-type');
      const title = document.getElementById('node-title-display');
      const desc = document.getElementById('node-desc-display');
      if (t) t.textContent = 'Проблема';
      if (title) title.textContent = 'Проблема обучения LLM принятию решений в OOD средах';
      if (desc) { desc.textContent = 'Агент не обобщает на новые ситуации в embodied-средах.'; desc.classList.remove('is-empty'); }
      document.getElementById('node-edit-block')?.classList.remove('hidden');
      document.getElementById('add-child-block')?.classList.add('hidden');
    }""")
    files.append(shot(page, f"{prefix}-07-node-edit"))

    reset_ui(page)
    set_view_mode(page, "researcher")
    if click_first_method_kanban(page):
        files.append(shot(page, f"{prefix}-08-kanban"))

        page.evaluate(
            """() => {
          const el = document.querySelector('.method-questions-trigger');
          if (el) el.dispatchEvent(new MouseEvent('click', { bubbles: true, cancelable: true }));
        }"""
        )
        page.wait_for_timeout(800)
        if page.locator("#method-questions-modal:not(.hidden)").count():
            files.append(shot(page, f"{prefix}-09-method-questions"))
            reset_ui(page)
            set_view_mode(page, "researcher")
            click_first_method_kanban(page)

        page.evaluate(
            """() => {
          const el = document.querySelector('.card-expand-report');
          if (el) el.dispatchEvent(new MouseEvent('click', { bubbles: true, cancelable: true }));
        }"""
        )
        page.wait_for_timeout(1200)
        if page.locator("#card-report-modal:not(.hidden)").count():
            files.append(shot(page, f"{prefix}-10-card-report"))
        reset_ui(page)

    reset_ui(page)
    set_view_mode(page, "researcher")
    page.locator("#btn-knowledge").click(force=True)
    page.wait_for_timeout(1200)
    files.append(shot(page, f"{prefix}-11-knowledge-base"))

    page.locator("#knowledge-tab-log").click(force=True)
    page.wait_for_timeout(400)
    files.append(shot(page, f"{prefix}-12-knowledge-log"))
    reset_ui(page)

    reset_ui(page)
    set_view_mode(page, "researcher")
    page.locator("#btn-paper").click(force=True)
    page.wait_for_timeout(1200)
    files.append(shot(page, f"{prefix}-13-paper-draft"))
    reset_ui(page)

    live = page.locator('button.method-activity-inspect, button[title*="Live"], button[title*="монитор"]').first
    if live.count():
        reset_ui(page)
        set_view_mode(page, "researcher")
        live.click(force=True)
        page.wait_for_timeout(800)
        files.append(shot(page, f"{prefix}-14-live-monitor"))

    return files


def capture_standalone(page, theme: str) -> list[str]:
    prefix = "light" if theme == "light" else "dark"
    files: list[str] = []

    page.goto(f"{BASE}/literature.html", wait_until="networkidle")
    set_theme(page, theme)
    page.reload(wait_until="networkidle")
    page.wait_for_timeout(1500)
    files.append(shot(page, f"{prefix}-15-literature"))

    page.goto(f"{BASE}/tour.html", wait_until="networkidle")
    set_theme(page, theme)
    page.reload(wait_until="networkidle")
    page.wait_for_timeout(1500)
    files.append(shot(page, f"{prefix}-16-tour"))

    return files


def main() -> None:
    arg = sys.argv[1] if len(sys.argv) > 1 else "both"
    themes = ["dark", "light"] if arg == "both" else (["light"] if arg == "light" else ["dark"])

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    manifest = {"capturedAt": datetime.now(timezone.utc).isoformat(), "base": BASE, "files": []}

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport=VIEWPORT, device_scale_factor=2)
        page = context.new_page()

        for theme in themes:
            manifest["files"].extend(capture_theme(page, theme))
            manifest["files"].extend(capture_standalone(page, theme))

        browser.close()

    (OUT_DIR / "manifest.json").write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    print(f"Captured {len(manifest['files'])} screenshots → {OUT_DIR}")
    for f in manifest["files"]:
        print(" ", Path(f).name)


if __name__ == "__main__":
    main()

File: scripts/capture-storyboard.mjs
#!/usr/bin/env node
/**
 * Capture ResearchOS UI states for Figma storyboard handoff.
 * Usage: node scripts/capture-storyboard.mjs [--theme dark|light|both]
 */
import { chromium } from "playwright";
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const OUT_DIR = path.join(ROOT, ".figma-storyboard");
const BASE = process.env.KOI_UI_URL || "http://127.0.0.1:8080";
const VIEWPORT = { width: 1440, height: 900 };

async function waitForApp(page) {
  await page.goto(`${BASE}/`, { waitUntil: "networkidle" });
  await page.waitForSelector("#project-list .project-list__item, #project-list button", {
    timeout: 60_000,
  });
  await page.waitForTimeout(1500);
}

async function resetUi(page) {
  await page.evaluate(() => {
    document.querySelectorAll(".modal").forEach((m) => m.classList.add("hidden"));
    document.getElementById("agent-chat-panel")?.classList.add("hidden");
    document.getElementById("rq-bell-panel")?.classList.add("hidden");
    document.getElementById("projects-sidebar")?.classList.remove("is-open");
    const toggle = document.getElementById("btn-projects-toggle");
    if (toggle) toggle.setAttribute("aria-expanded", "false");
  });
  await page.waitForTimeout(300);
}

async function setTheme(page, theme) {
  await page.evaluate((t) => {
    localStorage.setItem("koi-theme", t);
    document.documentElement.setAttribute("data-theme", t);
  }, theme);
  await page.waitForTimeout(200);
}

async function setViewMode(page, mode) {
  await page.selectOption("#view-mode-select", mode);
  await page.waitForTimeout(1200);
}

async function shot(page, name, opts = {}) {
  const file = path.join(OUT_DIR, `${name}.png`);
  await page.screenshot({ path: file, fullPage: false, ...opts });
  return file;
}

async function openSidebar(page) {
  const toggle = page.locator("#btn-projects-toggle");
  if ((await toggle.getAttribute("aria-expanded")) !== "true") {
    await toggle.click();
    await page.waitForTimeout(400);
  }
}

async function pickProjectWithData(page) {
  const llm = page.locator("#project-list button", {
    hasText: "Проблема обучения LLM",
  });
  if (await llm.count()) {
    await llm.first().click();
    await page.waitForTimeout(1200);
    return;
  }
  const any = page.locator("#project-list button[data-project-id]").first();
  await any.click();
  await page.waitForTimeout(1200);
}

async function clickFirstMethodKanban(page) {
  const btn = page.locator('button.node-kanban-below, button[class*="kanban"]').first();
  if (await btn.count()) {
    await btn.click({ force: true });
    await page.waitForTimeout(800);
    return true;
  }
  const method = page.locator('button[data-node-type="method"], .node-wrap[data-node-type="method"]').first();
  if (await method.count()) {
    await method.click({ force: true });
    await page.waitForTimeout(400);
  }
  const kanbanLink = page.locator('button:has-text("канбан"), button[title*="анбан"], .node-kanban-below').first();
  if (await kanbanLink.count()) {
    await kanbanLink.click({ force: true });
    await page.waitForTimeout(800);
    return true;
  }
  return false;
}

async function captureTheme(page, theme) {
  const prefix = theme === "light" ? "light" : "dark";
  await setTheme(page, theme);
  await resetUi(page);
  await waitForApp(page);
  await pickProjectWithData(page);

  const captures = [];

  // View modes
  for (const mode of ["chief", "teamlead", "researcher"]) {
    await resetUi(page);
    await setViewMode(page, mode);
    captures.push(await shot(page, `${prefix}-01-lab-${mode}`));
  }

  await resetUi(page);
  await setViewMode(page, "chief");
  await openSidebar(page);
  captures.push(await shot(page, `${prefix}-02-sidebar-open`));

  await resetUi(page);
  await setViewMode(page, "chief");
  await page.locator("#btn-agent-chat").click();
  await page.waitForTimeout(500);
  captures.push(await shot(page, `${prefix}-03-agent-chat`));

  await resetUi(page);
  await page.locator("#btn-rq-bell").click();
  await page.waitForTimeout(400);
  captures.push(await shot(page, `${prefix}-04-rq-bell`));

  await resetUi(page);
  await page.locator("#btn-settings").click();
  await page.waitForTimeout(400);
  captures.push(await shot(page, `${prefix}-05-settings`));

  await resetUi(page);
  await page.locator("#btn-new-project").click();
  await page.waitForTimeout(400);
  captures.push(await shot(page, `${prefix}-06-create-project`));

  await resetUi(page);
  await setViewMode(page, "researcher");
  const problem = page.locator('.node-wrap[data-node-type="problem"] button, button[data-node-type="problem"]').first();
  if (await problem.count()) {
    await problem.click({ force: true });
    await page.waitForTimeout(600);
    captures.push(await shot(page, `${prefix}-07-node-edit`));
  }

  await resetUi(page);
  await setViewMode(page, "researcher");
  if (await clickFirstMethodKanban(page)) {
    captures.push(await shot(page, `${prefix}-08-kanban`));

    const rqBtn = page.locator('button:has-text("Выводы:")').first();
    if (await rqBtn.count()) {
      await rqBtn.click({ force: true });
      await page.waitForTimeout(600);
      captures.push(await shot(page, `${prefix}-09-method-questions`));
      await page.keyboard.press("Escape");
      await page.waitForTimeout(300);
    }

    const reportBtn = page.locator('.kanban-card .card-report-link, .kanban-card button[title*="отч"]').first();
    if (await reportBtn.count()) {
      await reportBtn.click({ force: true });
      await page.waitForTimeout(600);
      captures.push(await shot(page, `${prefix}-10-card-report`));
      await page.locator('[data-close="card-report-modal"]').first().click();
      await page.waitForTimeout(300);
    }

    await page.locator('[data-close="kanban-modal"]').first().click();
    await page.waitForTimeout(300);
  }

  await resetUi(page);
  await setViewMode(page, "researcher");
  await page.locator("#btn-knowledge").click();
  await page.waitForTimeout(800);
  captures.push(await shot(page, `${prefix}-11-knowledge-base`));

  await page.locator("#knowledge-tab-log").click();
  await page.waitForTimeout(400);
  captures.push(await shot(page, `${prefix}-12-knowledge-log`));
  await page.locator('[data-close="knowledge-modal"]').first().click();

  await resetUi(page);
  await setViewMode(page, "researcher");
  await page.locator("#btn-paper").click();
  await page.waitForTimeout(800);
  captures.push(await shot(page, `${prefix}-13-paper-draft`));
  await page.locator('[data-close="paper-modal"]').first().click();

  const liveBtn = page.locator('button[title*="Live"], button[title*="монитор"], .method-activity-inspect').first();
  if (await liveBtn.count()) {
    await resetUi(page);
    await setViewMode(page, "researcher");
    await liveBtn.click({ force: true });
    await page.waitForTimeout(800);
    captures.push(await shot(page, `${prefix}-14-live-monitor`));
  }

  return captures;
}

async function captureStandalone(page, theme) {
  const prefix = theme === "light" ? "light" : "dark";
  const files = [];

  await setTheme(page, theme);

  await page.goto(`${BASE}/literature.html`, { waitUntil: "networkidle" });
  await page.waitForTimeout(1500);
  files.push(await shot(page, `${prefix}-15-literature`));

  await page.goto(`${BASE}/tour.html`, { waitUntil: "networkidle" });
  await page.waitForTimeout(1500);
  files.push(await shot(page, `${prefix}-16-tour`));

  return files;
}

async function main() {
  const arg = process.argv[2] || "both";
  const themes =
    arg === "both" ? ["dark", "light"] : arg === "light" ? ["light"] : ["dark"];

  await mkdir(OUT_DIR, { recursive: true });

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({ viewport: VIEWPORT, deviceScaleFactor: 2 });
  const page = await context.newPage();

  const manifest = { capturedAt: new Date().toISOString(), base: BASE, files: [] };

  for (const theme of themes) {
    manifest.files.push(...(await captureTheme(page, theme)));
    manifest.files.push(...(await captureStandalone(page, theme)));
  }

  await browser.close();

  await writeFile(
    path.join(OUT_DIR, "manifest.json"),
    JSON.stringify(manifest, null, 2),
    "utf8"
  );
  console.log(`Captured ${manifest.files.length} screenshots → ${OUT_DIR}`);
  for (const f of manifest.files) console.log(" ", path.basename(f));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
